### PR TITLE
Support traffic shaping for Kubelet CNI network plugin

### PR DESCRIPTION
**Pull request by m1093782566**
Currently the kubenet code supports applying basic traffic shaping during pod setup. This will happen if bandwidth-related annotations have been added to the pod's metadata, for example:

    {
        "kind": "Pod",
        "metadata": {
            "name": "iperf-slow",
            "annotations": {
                "kubernetes.io/ingress-bandwidth": "10M",
                "kubernetes.io/egress-bandwidth": "10M"
            }
        }
    }

Our current implementation uses the linux tc to add an download(ingress) and upload(egress) rate limiter using 1 root qdisc, 2 class(one for ingress and one for egress) and 2 filter(one for ingress and one for egress attached to the ingress and egress classes respectively).

Kubelet CNI code doesn't support it yet. However, there is enough functionality for us to add a simple CNI plugin to replicate the behavior we have today in kubenet if we feel this is an important feature. In fact, CNI has already supported traffic shaping plugin.

/sig network

### CODE REVIEW

**Comment by m1093782566** [comment]
/cc @cmluciano 

I guess you are interested in it? :)

File: contributors/design-proposals/network/support_traffic_shaping_for_kubelet_cni.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            "kubernetes.io/ingress-bandwidth": "10M",
+            "kubernetes.io/egress-bandwidth": "10M"
+        }
+    }
+}
+```
+
+Kubelet will parse the ingress and egress bandwidth values and transform them to Kb because both `ingressRate` and `egressRate` in cni bandwidth plugin are in Kbps. Eventually, kubelet will bulid the runtime config which like the following and call cni plugin `AddNetworkList` API.
+
+```json
```
Comment (caseydavenport) [inline contributors/design-proposals/network/support_traffic_shaping_for_kubelet_cni.md]: I've asked on the upstream plugin regarding support for runtimeConfig / capabilities similar to how the portmap plugin works today: https://github.com/containernetworking/plugins/pull/96#issuecomment-370952052

I think we want to agree a capability and use that to negotiate whether or not shaping will be done.

So a user would add something like this to their CNI config list if they wanted to enable traffic shaping via the plugin:

```json
{
  "type": "bandwidth",
  "capabilities": {"trafficShaping": true}
}
```

The kubelet would then populate the `runtimeConfig` section of the config when calling the `bandwidth` plugin:

```json
{
  "type": "bandwidth",
  "runtimeConfig": {
    "trafficShaping": {
      "ingressRate": "X",
      "egressRate": "Y"
    }
  }
}
```

**Comment by m1093782566** [comment]
Proposal updated as the CNI traffic shaping plugin got merged.

**Comment by Lion-Wei** [comment]
If we do this, then kubenet `BandwidthShaper` part can be replaced by traffic shaping plugin, big step of deprecating kubenet, I think.

**Comment by bowei** [comment]
@jingax10 can you take a look?

**Comment by m1093782566** [comment]
@jingax10 

> One quick question, can those ingress/egress limit be tuned later?

That's a good point. Technically, if `AddNetwork` and `AddNetworkList` of CNI APIs are idempotent, we can call these APIs as many times as required. So, ingress/egress limit be tuned via updating Pod annotations.

BTW, traffic shaping in kubenet only take effect in Pod setup.

**Comment by jingax10** [comment]
> That's a good point. Technically, if AddNetwork and AddNetworkList of CNI APIs are idempotent, we can call these APIs as many times as required. So, ingress/egress limit can be tuned via updating Pod annotations.
> 
> BTW, traffic shaping in kubenet ONLY takes effect in Pod setup.

Thanks for your reply! If limits are immutable, I could imagine it could be only used for bound the network usages from this pod if pod itself could become the bottleneck to serve (e.g., kubenet). Even if limits are mutable, how to set the limit dynamically is an art. Moreover, the granularity of the flows (for now only in and out without knowing the destinations) may not be sufficient for the control system to calculate the limits if one understands where are the bottlenecks. I am just trying to understand your use cases :-)

**Comment by thockin** [comment]
For the sake of my queue, I am unassigning myself.  Assign me (don't just CC me) when/if it is ready to be reviewed.

**Comment by caseydavenport** [comment]
I'm in favor of this as a way to move towards deprecating kubenet. The alternative would be to drop support for these annotations altogether, or revisit more holistically how we want to do bandwidth shaping (not sure how much appetite there is for such a discussion though).

Related, are these documented as alpha or beta somewhere or are these considered "GA"? My understanding was that these are an alpha type feature, but the annotation name doesn't indicate that.

**Comment by yujuhong** [comment]
> Related, are these documented as alpha or beta somewhere or are these considered "GA"?

Bandwidth shaping is an "experimental" feature, so I think the closest definition would be alpha.

> My understanding was that these are an alpha type feature, but the annotation name doesn't indicate that.

I believe the annotation was added before the alpha format was defined...

**Comment by m1093782566** [comment]
@caseydavenport 

This feature target alpha, I updated this document to clarify it.

@sig-network-leads any more comments? Can we push it forward?

**Comment by m1093782566** [comment]
/assign @thockin 

as it's ready for final review!

File: contributors/design-proposals/network/support_traffic_shaping_for_kubelet_cni.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+Kubelet would firstly parse the ingress and egress bandwidth values and transform them to Kbps because both `ingressRate` and `egressRate` in cni bandwidth plugin are in Kbps. Kubelet would  then detect whether user has enabled the traffic shaping plugin by checking the following CNI config file:
+
+```json
+{
+  "type": "bandwidth",
+  "capabilities": {"trafficShaping": true}
+}
+```
+
+If traffic shaping plugin is enabled, kubelet would populate the `runtimeConfig` section of the config when call the `bandwidth` plugin:
```
Comment (thockin) [inline contributors/design-proposals/network/support_traffic_shaping_for_kubelet_cni.md]: Can we just always pass this, so no checking required?

**Comment by thockin** [comment]
This is OK for now.  We need a long-term answer that respects the general resource model and plugin model.

/lgtm
/approve

**Comment by m1093782566** [comment]
>  We need a long-term answer that respects the general resource model and plugin model.

Yeah, definitely. That's also what my team is exploring :)

**Pull request merged**